### PR TITLE
[3.x] Create Runtime CFN template for Custom CLI Credentails

### PR DESCRIPTION
**Pull request by himani2411**
### Description of changes
* Creating custom CLI credentials to run Resource prefix integration test case through a runtime change in existing CFN template user-role.cfn.yaml
* Removing template [user-role-iam-resource-prefix.cfn.yaml](https://github.com/aws/aws-parallelcluster/compare/develop...himani2411:aws-parallelcluster:wip/runtime-iam-role#diff-6c891a2944891d4a7fea672dba07bc0ef430cf2e316f196359616003b04d3123) which was used for creating CLI credentials.

### Tests
* Running the test_iam_resource_prefix test case locally
* Running jenkins pcluster3 test cases

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/iam_policies/user-role.cfn.yaml [hunk 1]
```diff
@@ -823,10 +823,12 @@ Resources:
               - dynamodb:BatchWriteItem
             Resource: !Sub arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/parallelcluster-*
             Effect: Allow
-          - Action: ec2:DescribeInstances
```
Comment (dreambeyondorange) [inline tests/iam_policies/user-role.cfn.yaml:826]: Why do we need to convert all these to lists?  Does this affect current tests?

File: tests/integration-tests/tests/iam/test_iam.py [hunk 2]
```diff
@@ -363,6 +369,233 @@ def test_iam_resource_prefix(
             _test_iam_resource_in_cluster(cfn_client, iam_client, cluster.name, iam_resource_prefix)
 
 
+def _remove_from_config(config, item_list):
```
Comment (dreambeyondorange) [inline tests/integration-tests/tests/iam/test_iam.py]: Seems like you are building some general purpose utilities to update the config files.  Can you add descriptions and explanations for the functions you've created here?  Its a bit hard to follow how they all fit together.

File: tests/iam_policies/user-role.cfn.yaml (hunk 1 above)
Comment (himani2411) [reply tests/iam_policies/user-role.cfn.yaml:826]: No this doesnt affect the test. I just converted it to a list so that when I am trying to parse this through a template generator and make my changes, I can use a generic functionality rather than creating another function to handle it as all the other policy statements have been a list so far.

Comment (himani2411) [reply tests/iam_policies/user-role.cfn.yaml:826]: I hav already tested the changes using jenkins

File: tests/integration-tests/tests/iam/test_iam.py (hunk 2 above)
Comment (himani2411) [reply tests/integration-tests/tests/iam/test_iam.py]: Sure.

File: tests/integration-tests/tests/iam/test_iam.py [hunk 3]
```diff
@@ -363,6 +369,247 @@ def test_iam_resource_prefix(
...
+    """Remove a list of sections from Config template."""
+    if isinstance(config, dict):
+        for item in section_list:
+            config.pop(item)
+    elif isinstance(config, list):
+        for item in section_list:
+            config.remove(item)
+
+
+def _replace_resource(config_string, list_of_dict):
```
Comment (dreambeyondorange) [inline tests/integration-tests/tests/iam/test_iam.py:382]: Is this replacing all the different substrings with one?

File: tests/integration-tests/tests/iam/test_iam.py [hunk 4]
```diff
@@ -363,6 +369,247 @@ def test_iam_resource_prefix(
             _test_iam_resource_in_cluster(cfn_client, iam_client, cluster.name, iam_resource_prefix)
 
 
+def _remove_from_config(config, section_list):
```
Comment (dreambeyondorange) [inline tests/integration-tests/tests/iam/test_iam.py:372]: It seems like you are doing the replacement manually.  Why not use a tool like jinja for this?

File: tests/integration-tests/tests/iam/test_iam.py [hunk 5]
```diff
@@ -363,6 +369,247 @@ def test_iam_resource_prefix(
...
+        ):
+            permission_boundary_statements.remove(statement)
+
+
+def _find_resource_action(resource_type, action_list):
+    """Check if a specific resource type exists in the Policy statement's Actions."""
+    return any(resource_type in action for action in action_list)
+
+
+def _create_iam_user_role_config(cluster_config, iam_resource_prefix):
```
Comment (gmarciani) [inline tests/integration-tests/tests/iam/test_iam.py:542]: #### Major
1. In general, we should always  use Jinja to make runtime changes to a cloudformation template instead of ad-hoc "manual changes".
2. We can consider the following alternative approach instead of making direct changes to the user role template: implement a fixture that creates a permissions boundary policy according to the given IAM prefix and applies it to the user role

#### Minor
The value returned by this function is a stack template, not a generic config. I suggest to rename this function "_create_iam_user_role_template"

**Review by gmarciani — CHANGES_REQUESTED** [review]

File: tests/integration-tests/tests/iam/test_iam.py (hunk 5 above)
Comment (himani2411) [reply tests/integration-tests/tests/iam/test_iam.py:542]: Creating a new [PR](https://github.com/aws/aws-parallelcluster/pull/4782) to address new design changes
